tests: Add the description for vrf integration tests

liangwen12year opened the pull request:
In the vrf integration tests, apply the desired state with description, so that automatically generated state examples can be used for nmstate-yamlsmith model training.

liangwen12year committed:
tests: Add the description for vrf integration tests

In the vrf integration tests, apply the desired state with description,
so that automatically generated state examples can be used for
nmstate-yamlsmith model training.

Signed-off-by: Wen Liang <liangwen12year@gmail.com>

liangwen12year commented:
```
description: Configure the vrf interface test-vrf0 with vrf port eth1
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port:
    - eth1
    route-table-id: 0
description: Create the vrf interface test-vrf0 with vrf port eth1 and vrf route table ID 100
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port:
    - eth1
    route-table-id: 0
description: Delete the vrf interface test-vrf0
interfaces:
- name: test-vrf0
  type: vrf
  state: absent
description: Create the vrf interface test-vrf0 with vrf port eth1 and eth2 attached
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port:
    - eth1
    - eth2
    route-table-id: 0
description: Configure the vrf device test-vrf0 with empty port and vrf route table ID 101
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port: null
    route-table-id: 101
description: Delete the vrf interface test-vrf0
interfaces:
- name: test-vrf0
  type: vrf
  state: absent
description: Create the vrf interface test-vrf0 with vrf port eth1 and vrf route table ID 100
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port:
    - eth1
    route-table-id: 100
description: Create the vrf interface test-vrf0 with vrf port eth1 and vrf route table ID 100
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port:
    - eth1
    route-table-id: 100
description: Create the test-vrf0 with port eth1 attached
interfaces: []
description: Delete the vrf interface test-vrf0
interfaces:
- name: test-vrf0
  type: vrf
  state: absent
description: Create the vrf interface test-vrf0 with vrf port eth1 and vrf route table ID 100
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port:
    - eth1
    route-table-id: 100
description: Delete the vrf interface test-vrf0
interfaces:
- name: test-vrf0
  type: vrf
  state: absent
description: Create the vrf interface test-vrf0 with the empty ports and the vrf route table ID 100
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port: []
    route-table-id: 100
description: Create the vrf interface test-vrf0 with the empty port, create the vrf interface test-vrf1 with the vrf port eth1 and eth2 attached
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port: []
    route-table-id: 0
- name: test-vrf1
  type: vrf
  state: up
  vrf:
    port:
    - eth1
    - eth2
    route-table-id: 0
description: Delete the vrf interface test-vrf0
interfaces:
- name: test-vrf0
  type: vrf
  state: absent
description: Delete the vrf interface test-vrf1
interfaces:
- name: test-vrf1
  type: vrf
  state: absent
description: Create the vrf interface test-vrf0 with vrf port eth1 and vrf route table ID 100
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port:
    - eth1
    route-table-id: 100
description: Create the vrf interface test-vrf1 with vrf port eth2 and vrf route table ID 101
interfaces:
- name: test-vrf1
  type: vrf
  state: up
  vrf:
    port:
    - eth2
    route-table-id: 101
description: Delete the vlan interface test-bond0.100, delete the bond interface test-bond0, delete the vrf interface test-vrf0
interfaces:
- name: test-bond0
  type: bond
  state: absent
- name: test-bond0.100
  type: vlan
  state: absent
- name: test-vrf0
  type: vrf
  state: absent
description: Create the vlan device with the base interface test-bond0 and vlan id 100, create the bond interface test-bond0 with bonding mode balance-rr, create the vrf interface test-vrf0 with ports test-bond0, test-bond0.100, and the vrf route table id 100
interfaces:
- name: test-bond0
  type: bond
  state: up
  controller: test-vrf0
  link-aggregation:
    mode: balance-rr
- name: test-bond0.100
  type: vlan
  state: up
  controller: test-vrf0
  vlan:
    base-iface: test-bond0
    id: 100
    reorder-headers: true
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port:
    - test-bond0
    - test-bond0.100
    route-table-id: 100
description: Delete the vrf interface test-vrf0
interfaces:
- name: test-vrf0
  type: vrf
  state: absent
description: Create the vrf interface test-vrf0 with vrf port eth1 and vrf route table ID 100
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port:
    - eth1
    route-table-id: 100
description: Set the empty ports for the vrf interface test-vrf0
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port: []
    route-table-id: 0
description: Delete the vrf interface test-vrf0
interfaces:
- name: test-vrf0
  type: vrf
  state: absent
description: Create the vrf interface test-vrf0 with vrf port eth1 and vrf route table ID 100
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port:
    - eth1
    route-table-id: 100
description: Create the vrf interface test-vrf0 with vrf port eth1 and vrf route table ID 100
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port:
    - eth1
    route-table-id: 100
description: Delete the vrf interface test-vrf0
interfaces:
- name: test-vrf0
  type: vrf
  state: absent
description: Set the empty ports for the vrf interface test-vrf0
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port: []
    route-table-id: 0
description: Create the vrf interface test-vrf1, attach ports eth1 and eth2 to it, configure the vrf route table ID 101
interfaces:
- name: test-vrf1
  type: vrf
  state: up
  vrf:
    port:
    - eth1
    - eth2
    route-table-id: 101
description: Delete the vrf interface test-vrf1
interfaces:
- name: test-vrf1
  type: vrf
  state: absent
description: Create the vrf interface test-vrf0 with port eth2, create the vrf interface test-vrf1 with port eth1
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port:
    - eth2
    route-table-id: 0
- name: test-vrf1
  type: vrf
  state: up
  vrf:
    port:
    - eth1
    route-table-id: 0
description: Create the vrf interface test-vrf1 with vrf port eth2 and vrf route table ID 101
interfaces:
- name: test-vrf1
  type: vrf
  state: up
  vrf:
    port:
    - eth2
    route-table-id: 101
description: Delete the vrf interface test-vrf1
interfaces:
- name: test-vrf1
  type: vrf
  state: absent
description: Delete the vrf interface test-vrf0
interfaces:
- name: test-vrf0
  type: vrf
  state: absent
description: Create the vrf interface test-vrf0 with vrf port eth1 and vrf route table ID 100
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port:
    - eth1
    route-table-id: 100
description: Delete the test-vrf1 interface
interfaces:
- name: test-vrf1
  type: vrf
  state: absent
description: Delete the vrf interface test-vrf0 and veth device veth0
interfaces:
- name: test-vrf0
  type: vrf
  state: absent
- name: veth0
  type: ethernet
  state: absent
description: Configure the vrf interface test-vrf1 with ports veth0 and vrf route table ID 101, configure the veth interface veth0 with controller test-vrf1
interfaces:
- name: test-vrf1
  type: vrf
  state: up
  vrf:
    port:
    - veth0
    route-table-id: 101
- name: veth0
  type: ethernet
  state: up
  controller: test-vrf1
description: Delete the vrf interface test-vrf0
interfaces:
- name: test-vrf0
  type: vrf
  state: absent
description: Create the vrf interface test-vrf0 with vrf port eth1 and vrf route table ID 100
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port:
    - eth1
    route-table-id: 100
description: Delete the vrf interface test-vrf0
interfaces:
- name: test-vrf0
  type: vrf
  state: absent
description: Create the vrf interface test-vrf0 with vrf port eth1 and vrf route table ID 100
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port:
    - eth1
    route-table-id: 100
description: Create the vlan device with the base interface test-bond0 and vlan id 100, create the bond interface test-bond0 with bonding mode balance-rr, create the vrf interface test-vrf0 with ports test-bond0, test-bond0.100, and the vrf route table id 100
interfaces:
- name: test-bond0
  type: bond
  state: up
  controller: test-vrf0
  link-aggregation:
    mode: balance-rr
- name: test-bond0.100
  type: vlan
  state: up
  controller: test-vrf0
  vlan:
    base-iface: test-bond0
    id: 100
    reorder-headers: true
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port:
    - test-bond0
    - test-bond0.100
    route-table-id: 100
description: Remove the vrf device test-vrf0 and the bond device test-bond0
interfaces:
- name: test-bond0
  type: bond
  state: absent
- name: test-vrf0
  type: vrf
  state: absent
description: Delete the vlan interface test-bond0.100, delete the bond interface test-bond0, delete the vrf interface test-vrf0
interfaces:
- name: test-bond0
  type: bond
  state: absent
- name: test-bond0.100
  type: vlan
  state: absent
- name: test-vrf0
  type: vrf
  state: absent
```

cathay4t reviewed (changes requested):
The test code never use table ID 0, we need investigation why it happens.

cathay4t commented:
Never mind. Let me fix the vrf gen_diff.

cathay4t commented:
https://github.com/nmstate/nmstate/pull/2759 will fix the VRF `route-table-id: 0` issue.

liangwen12year commented:
> #2759 will fix the VRF `route-table-id: 0` issue.

Thanks for fixing the `gen_diff`, I updated this PR.

cathay4t commented on tests/integration/vrf_test.yaml:
```diff
@@ -0,0 +1,368 @@
+    port:
+    - eth1
+    route-table-id: 100
+description: Configure the vrf device test-vrf0 with empty port and vrf route table ID 101
```
There is no empty port for this YAML.

cathay4t commented on tests/integration/vrf_test.yaml:
```diff
@@ -0,0 +1,368 @@
+    port:
+    - eth1
+    - eth2
+description: Create the vrf interface test-vrf0 with vrf port eth1 and vrf route table ID 100
```
There is not table ID for this YAML.

liangwen12year replied:
The diff state is generated by the test `test_add_and_remove_port`, where the vrf table id 100 is indeed specified, however, the previous test is `test_create_with_empty_ports`, and the vrf table id 100 is applied but there is no clean up in that function.

liangwen12year replied:
I added a commit to this PR for fixing the issue.

cathay4t commented on tests/integration/vrf_test.yaml:
```diff
@@ -0,0 +1,368 @@
+    port:
+    - eth1
+    route-table-id: 100
+description: Set the empty ports and route table id 0 for the vrf interface test-vrf0
```
There is no table ID 0 in this YAML.

cathay4t commented on tests/integration/vrf_test.yaml:
```diff
@@ -0,0 +1,368 @@
+  state: up
+  vrf:
+    port: []
+description: Set the empty ports and route table id 0 for the vrf interface test-vrf0
```
There is no table ID 0 in this YAML.

liangwen12year committed:
vrf: Clean up the vrf device at the end of test

`test_create_with_empty_ports` also applies the vrf route table ID 100
in the desired state, if we do not clean up the vrf interface properly,
the diff state in the following up test `test_add_and_remove_port` will
not contain the vrf route table ID 100.

Signed-off-by: Wen Liang <liangwen12year@gmail.com>

cathay4t commented on tests/integration/vrf_test.yaml:
```diff
@@ -0,0 +1,369 @@
+  state: up
+  vrf:
+    port:
+    - eth2
```
Even it is hard to tell whether certain YAML is creating or modifying, but clearly you cannot create a VRF without route table ID defined. So this YAML is modifying the vrf interface, not create.

cathay4t commented on tests/integration/vrf_test.yaml:
```diff
@@ -0,0 +1,369 @@
+- name: veth0
+  type: ethernet
+  state: absent
+description: Configure the vrf interface test-vrf1 with ports veth0 and vrf route table ID 101, configure the veth interface veth0 with controller test-vrf1
```
Normal user does not say `config xxx with controller`, but might use `attach ethernet veth0 to test-vrf1`

liangwen12year commented:
```
description: Attach ethernet eth1 to the vrf interface test-vrf0
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port:
    - eth1
description: Delete the vrf interface test-vrf0
interfaces:
- name: test-vrf0
  type: vrf
  state: absent
description: Create the vrf interface test-vrf0 with vrf port eth1 and vrf route table ID 100
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port:
    - eth1
    route-table-id: 100
description: Attach ethernet eth1 and eth2 to the vrf interface test-vrf0
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port:
    - eth1
    - eth2
description: Configure the vrf device test-vrf0 with vrf route table ID 101
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    route-table-id: 101
description: Delete the vrf interface test-vrf0
interfaces:
- name: test-vrf0
  type: vrf
  state: absent
description: Create the vrf interface test-vrf0 with vrf port eth1 and vrf route table ID 100
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port:
    - eth1
    route-table-id: 100
description: Delete the vrf interface test-vrf0
interfaces:
- name: test-vrf0
  type: vrf
  state: absent
description: Create the vrf interface test-vrf0 with vrf port eth1 and vrf route table ID 100
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port:
    - eth1
    route-table-id: 100
description: Delete the vrf interface test-vrf0
interfaces:
- name: test-vrf0
  type: vrf
  state: absent
description: Create the vrf interface test-vrf0 with vrf port eth1 and vrf route table ID 100
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port:
    - eth1
    route-table-id: 100
description: Delete the vrf interface test-vrf0
interfaces:
- name: test-vrf0
  type: vrf
  state: absent
description: Create the vrf interface test-vrf0 with the empty ports and the vrf route table ID 100
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port: []
    route-table-id: 100
description: Configure the vrf interface test-vrf0 with empty port, configure the vrf interface test-vrf1 with the vrf port eth1 and eth2
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port: []
- name: test-vrf1
  type: vrf
  state: up
  vrf:
    port:
    - eth1
    - eth2
description: Delete the vrf interface test-vrf1
interfaces:
- name: test-vrf1
  type: vrf
  state: absent
description: Create the vrf interface test-vrf1 with vrf port eth2 and vrf route table ID 101
interfaces:
- name: test-vrf1
  type: vrf
  state: up
  vrf:
    port:
    - eth2
    route-table-id: 101
description: Delete the vrf interface test-vrf0
interfaces:
- name: test-vrf0
  type: vrf
  state: absent
description: Create the vrf interface test-vrf0 with vrf port eth1 and vrf route table ID 100
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port:
    - eth1
    route-table-id: 100
description: Delete the vlan interface test-bond0.100, delete the bond interface test-bond0, delete the vrf interface test-vrf0
interfaces:
- name: test-bond0
  type: bond
  state: absent
- name: test-bond0.100
  type: vlan
  state: absent
- name: test-vrf0
  type: vrf
  state: absent
description: Create the vlan device over test-bond0 with ID 100, create the bond interface test-bond0 with bonding mode balance-rr, create the vrf interface test-vrf0 with ports test-bond0, test-bond0.100, and the vrf route table id 100
interfaces:
- name: test-bond0
  type: bond
  state: up
  link-aggregation:
    mode: balance-rr
- name: test-bond0.100
  type: vlan
  state: up
  vlan:
    base-iface: test-bond0
    id: 100
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port:
    - test-bond0
    - test-bond0.100
    route-table-id: 100
description: Delete the vrf interface test-vrf0
interfaces:
- name: test-vrf0
  type: vrf
  state: absent
description: Create the vrf interface test-vrf0 with vrf port eth1 and vrf route table ID 100
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port:
    - eth1
    route-table-id: 100
description: Create the vrf interface test-vrf0 with vrf port eth1 and vrf route table ID 100
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port:
    - eth1
    route-table-id: 100
description: Delete the vrf interface test-vrf0
interfaces:
- name: test-vrf0
  type: vrf
  state: absent
description: Set the empty port for the vrf interface test-vrf0
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port: []
description: Create the vrf interface test-vrf0 with vrf port eth1 and vrf route table ID 100
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port:
    - eth1
    route-table-id: 100
description: Delete the vrf interface test-vrf0
interfaces:
- name: test-vrf0
  type: vrf
  state: absent
description: Set the empty port for the vrf interface test-vrf0
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port: []
description: Create the vrf interface test-vrf1, attach ports eth1 and eth2 to it, configure the vrf route table ID 101
interfaces:
- name: test-vrf1
  type: vrf
  state: up
  vrf:
    port:
    - eth1
    - eth2
    route-table-id: 101
description: Delete the vrf interface test-vrf1
interfaces:
- name: test-vrf1
  type: vrf
  state: absent
description: Create the vrf interface test-vrf1 with vrf port eth2 and vrf route table ID 101
interfaces:
- name: test-vrf1
  type: vrf
  state: up
  vrf:
    port:
    - eth2
    route-table-id: 101
description: Attach ethernet eth2 to the vrf interface test-vrf0, attach ethernet eth1 to the vrf interface test-vrf1
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port:
    - eth2
- name: test-vrf1
  type: vrf
  state: up
  vrf:
    port:
    - eth1
description: Delete the vrf interface test-vrf0
interfaces:
- name: test-vrf0
  type: vrf
  state: absent
description: Delete the vrf interface test-vrf1
interfaces:
- name: test-vrf1
  type: vrf
  state: absent
description: Create the vrf interface test-vrf0 with vrf port eth1 and vrf route table ID 100
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port:
    - eth1
    route-table-id: 100
description: Delete the vrf interface test-vrf0 and veth device veth0
interfaces:
- name: test-vrf0
  type: vrf
  state: absent
- name: veth0
  type: ethernet
  state: absent
description: Delete the test-vrf1 interface
interfaces:
- name: test-vrf1
  type: vrf
  state: absent
description: Attach ethernet veth0 to the vrf interface test-vrf1 and set vrf route table ID to 101
interfaces:
- name: test-vrf1
  type: vrf
  state: up
  vrf:
    port:
    - veth0
    route-table-id: 101
- name: veth0
  type: ethernet
  state: up
  controller: test-vrf1
description: Delete the vrf interface test-vrf0
interfaces:
- name: test-vrf0
  type: vrf
  state: absent
description: Create the vrf interface test-vrf0 with vrf port eth1 and vrf route table ID 100
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port:
    - eth1
    route-table-id: 100
description: Delete the vrf interface test-vrf0
interfaces:
- name: test-vrf0
  type: vrf
  state: absent
description: Create the vrf interface test-vrf0 with vrf port eth1 and vrf route table ID 100
interfaces:
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port:
    - eth1
    route-table-id: 100
description: Remove the vrf device test-vrf0 and the bond device test-bond0
interfaces:
- name: test-bond0
  type: bond
  state: absent
- name: test-vrf0
  type: vrf
  state: absent
description: Delete the vlan interface test-bond0.100, delete the bond interface test-bond0, delete the vrf interface test-vrf0
interfaces:
- name: test-bond0
  type: bond
  state: absent
- name: test-bond0.100
  type: vlan
  state: absent
- name: test-vrf0
  type: vrf
  state: absent
description: Create the vlan device over test-bond0 with ID 100, create the bond interface test-bond0 with bonding mode balance-rr, create the vrf interface test-vrf0 with ports test-bond0, test-bond0.100, and the vrf route table id 100
interfaces:
- name: test-bond0
  type: bond
  state: up
  link-aggregation:
    mode: balance-rr
- name: test-bond0.100
  type: vlan
  state: up
  vlan:
    base-iface: test-bond0
    id: 100
- name: test-vrf0
  type: vrf
  state: up
  vrf:
    port:
    - test-bond0
    - test-bond0.100
    route-table-id: 100
```
